File: scripts/testing/dataset_upsert_smoke.py
"""Smoke test the full dataset upsert/update functionality (Python).

Three-step flow via ``client.datasets.create_dataset`` (which sends
``action=update`` under the hood) using the per-example ``id`` field for
stable identity matching:

  1. seed       — 4 examples (a, b, c, d) → new version, 4 examples
  2. idempotence — re-send same payload   → same version_id, no-op
  3. modify     — patch a (A → A'), drop b, move c's split, keep d, add e
                   → new version, 4 examples
                   → because each example carries an `id`, the diff
                     classifies a as a PATCH revision (not delete+create),
                     b as DELETE, e as CREATE, and c as a splits-only
                     change (no new revision row)

Run:
    uv run scripts/testing/dataset_upsert_smoke.py

Set PHOENIX_COLLECTOR_ENDPOINT (e.g. http://localhost:6112) and/or
PHOENIX_API_KEY if your server isn't on the default http://localhost:6006
with no auth.
"""

import time

from phoenix.client import Client


def _run_experiment_and_assert(
    client: Client,
    dataset,
    *,
    label: str,
    expected_outputs: set[str],
) -> None:
    result = client.experiments.run_experiment(
        dataset=dataset,
        task=lambda input: input["v"],
        experiment_name=f"{dataset.name}-{label}-experiment",
        print_summary=False,
    )
    actual_outputs = {str(run["output"]) for run in result["task_runs"]}
    print(
        f"{label} experiment: experiment_id={result['experiment_id']} "
        f"version={dataset.version_id} runs={len(result['task_runs'])}"
    )
    assert len(result["task_runs"]) == len(dataset)
    assert len(result["evaluation_runs"]) == 0
    assert actual_outputs == expected_outputs, (
        f"expected experiment outputs {expected_outputs}, got {actual_outputs}"
    )


def main() -> None:
    client = Client()
    name = f"upsert-smoke-{int(time.time())}"

    # ── step 1: seed ─────────────────────────────────────────────────────
    seed = [
        {"id": "a", "input": {"v": "A"}, "output": {}, "splits": "s1"},
        {"id": "b", "input": {"v": "B"}, "output": {}, "splits": "s1"},
        {"id": "c", "input": {"v": "C"}, "output": {}, "splits": "s2"},
        {"id": "d", "input": {"v": "D"}, "output": {}, "splits": "s2"},
    ]
    d1 = client.datasets.create_dataset(name=name, examples=seed)
    print(f"step 1 seed         : version={d1.version_id}  n={len(d1)}")
    assert len(d1) == 4
    _run_experiment_and_assert(
        client,
        d1,
        label="step-1-seed",
        expected_outputs={"A", "B", "C", "D"},
    )

    # ── step 2: idempotence ──────────────────────────────────────────────
    d2 = client.datasets.create_dataset(name=name, examples=seed)
    print(f"step 2 idempotence  : version={d2.version_id}  n={len(d2)}")
    assert d2.version_id == d1.version_id, "expected no new version"
    assert len(d2) == 4
    _run_experiment_and_assert(
        client,
        d2,
        label="step-2-idempotence",
        expected_outputs={"A", "B", "C", "D"},
    )

    # ── step 3: modify ───────────────────────────────────────────────────
    # a: A → A'  (PATCH, matched by id="a")
    # b: dropped (DELETE)
    # c: same content, split s2 → s1 (no revision, splits-only change)
    # d: unchanged
    # e: new (CREATE)
    modified = [
        {"id": "a", "input": {"v": "A'"}, "output": {}, "splits": "s1"},
        {"id": "c", "input": {"v": "C"}, "output": {}, "splits": "s1"},
        {"id": "d", "input": {"v": "D"}, "output": {}, "splits": "s2"},
        {"id": "e", "input": {"v": "E"}, "output": {}, "splits": "s1"},
    ]
    d3 = client.datasets.create_dataset(name=name, examples=modified)
    print(f"step 3 modify       : version={d3.version_id}  n={len(d3)}")
    assert d3.version_id != d1.version_id, "expected new version"
    assert len(d3) == 4
    _run_experiment_and_assert(
        client,
        d3,
        label="step-3-modify",
        expected_outputs={"A'", "C", "D", "E"},
    )

    inputs_now = [dict(ex["input"]) for ex in d3]
    assert {"v": "A'"} in inputs_now and {"v": "E"} in inputs_now
    assert {"v": "B"} not in inputs_now

    # verify split-only change on c: it should now be in split s1
    s1_after = client.datasets.get_dataset(dataset=name, splits=["s1"])
    s1_inputs = [dict(ex["input"]) for ex in s1_after]
    assert {"v": "C"} in s1_inputs, (
        f"expected example c (input {{'v':'C'}}) to be in split s1; got {s1_inputs}"
    )

    print(f"OK — full upsert flow verified (dataset {name!r})")


if __name__ == "__main__":
    main()

File: scripts/testing/dataset_upsert_smoke.ts
/* eslint-disable no-console */
/**
 * Smoke test the full dataset upsert/update functionality (TypeScript).
 *
 * Three-step flow via `createDataset` (which sends `action=update` under the
 * hood) using the per-example `id` field for stable identity matching:
 *
 *   1. seed        - 4 examples (a, b, c, d) -> new version, 4 examples
 *   2. idempotence - re-send same payload    -> same versionId, no-op
 *   3. modify      - patch a (A -> A'), drop b, move c's split, keep d, add e
 *                    -> new version, 4 examples
 *                    -> because each example carries an `id`, the diff
 *                       classifies a as a PATCH revision (not delete+create),
 *                       b as DELETE, e as CREATE, and c as a splits-only
 *                       change (no new revision row)
 *
 * Run:
 *   pnpm --dir js --filter @arizeai/phoenix-client exec tsx ../../../scripts/testing/dataset_upsert_smoke.ts
 *
 * Set PHOENIX_HOST (for example http://localhost:6148) and/or PHOENIX_API_KEY
 * if your server is not on the default http://localhost:6006 with no auth.
 */

import { createClient } from "../../js/packages/phoenix-client/src/client";
import {
  createDataset,
  getDataset,
} from "../../js/packages/phoenix-client/src/datasets";
import { runExperiment } from "../../js/packages/phoenix-client/src/experiments";
import type {
  Dataset,
  Example,
} from "../../js/packages/phoenix-client/src/types/datasets";

type SmokeExample = Example & { id: string };

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

async function runExperimentAndAssert({
  dataset,
  label,
  expectedOutputs,
}: {
  dataset: Dataset;
  label: string;
  expectedOutputs: Set<string>;
}) {
  const experiment = await runExperiment({
    client,
    dataset: { datasetId: dataset.id, versionId: dataset.versionId },
    experimentName: `${dataset.name}-${label}-experiment`,
    task: async (example) => String(example.input.v),
    logger: console,
  });
  const taskRuns = Object.values(experiment.runs);
  const actualOutputs = new Set(taskRuns.map((taskRun) => String(taskRun.output)));

  console.log(
    [
      `${label} experiment:`,
      `experiment_id=${experiment.id}`,
      `version=${dataset.versionId}`,
      `runs=${taskRuns.length}`,
    ].join(" ")
  );

  assert(
    taskRuns.length === dataset.examples.length,
    `expected ${dataset.examples.length} task runs, got ${taskRuns.length}`
  );
  assert(
    (experiment.evaluationRuns?.length ?? 0) === 0,
    `expected no evaluation runs, got ${experiment.evaluationRuns?.length ?? 0}`
  );
  assert(
    actualOutputs.size === expectedOutputs.size &&
      [...expectedOutputs].every((expectedOutput) => actualOutputs.has(expectedOutput)),
    `expected experiment outputs ${JSON.stringify(
      [...expectedOutputs]
    )}, got ${JSON.stringify([...actualOutputs])}`
  );
}

async function main() {
  const name = `upsert-smoke-ts-${Date.now()}`;
  const description = "Smoke test for dataset upsert semantics";

  const seed: SmokeExample[] = [
    { id: "a", input: { v: "A" }, output: {}, splits: "s1" },
    { id: "b", input: { v: "B" }, output: {}, splits: "s1" },
    { id: "c", input: { v: "C" }, output: {}, splits: "s2" },
    { id: "d", input: { v: "D" }, output: {}, splits: "s2" },
  ];

  const step1 = await createDataset({
    client,
    name,
    description,
    examples: seed,
  });
  const dataset1 = await getDataset({
    client,
    dataset: { datasetId: step1.datasetId },
  });
  console.log(`step 1 seed         : version=${dataset1.versionId}  n=${dataset1.examples.length}`);
  assert(dataset1.examples.length === 4, `expected 4 examples, got ${dataset1.examples.length}`);
  await runExperimentAndAssert({
    dataset: dataset1,
    label: "step-1-seed",
    expectedOutputs: new Set(["A", "B", "C", "D"]),
  });

  const step2 = await createDataset({
    client,
    name,
    description,
    examples: seed,
  });
  const dataset2 = await getDataset({
    client,
    dataset: { datasetId: step2.datasetId },
  });
  console.log(`step 2 idempotence  : version=${dataset2.versionId}  n=${dataset2.examples.length}`);
  assert(
    dataset2.versionId === dataset1.versionId,
    `expected no new version, got ${dataset1.versionId} -> ${dataset2.versionId}`
  );
  assert(dataset2.examples.length === 4, `expected 4 examples, got ${dataset2.examples.length}`);
  await runExperimentAndAssert({
    dataset: dataset2,
    label: "step-2-idempotence",
    expectedOutputs: new Set(["A", "B", "C", "D"]),
  });

  const modified: SmokeExample[] = [
    { id: "a", input: { v: "A'" }, output: {}, splits: "s1" },
    { id: "c", input: { v: "C" }, output: {}, splits: "s1" },
    { id: "d", input: { v: "D" }, output: {}, splits: "s2" },
    { id: "e", input: { v: "E" }, output: {}, splits: "s1" },
  ];

  const step3 = await createDataset({
    client,
    name,
    description,
    examples: modified,
  });
  const dataset3 = await getDataset({
    client,
    dataset: { datasetId: step3.datasetId },
  });
  console.log(`step 3 modify       : version=${dataset3.versionId}  n=${dataset3.examples.length}`);
  assert(
    dataset3.versionId !== dataset1.versionId,
    `expected new version, got ${dataset1.versionId}`
  );
  assert(dataset3.examples.length === 4, `expected 4 examples, got ${dataset3.examples.length}`);
  await runExperimentAndAssert({
    dataset: dataset3,
    label: "step-3-modify",
    expectedOutputs: new Set(["A'", "C", "D", "E"]),
  });

  const currentInputs = dataset3.examples.map((example) => example.input);
  const hasPatchedExample = currentInputs.some((input) => input.v === "A'");
  const hasNewExample = currentInputs.some((input) => input.v === "E");
  const hasDeletedExample = currentInputs.some((input) => input.v === "B");
  assert(hasPatchedExample, "expected patched example A' in latest dataset");
  assert(hasNewExample, "expected new example E in latest dataset");
  assert(!hasDeletedExample, "did not expect deleted example B in latest dataset");

  const s1After = await getDataset({
    client,
    dataset: { datasetName: name, splits: ["s1"] },
  });
  const s1Inputs = s1After.examples.map((example) => example.input);
  const hasMovedExample = s1Inputs.some((input) => input.v === "C");
  assert(
    hasMovedExample,
    `expected example c (input {"v":"C"}) to be in split s1; got ${JSON.stringify(s1Inputs)}`
  );

  console.log(`OK - full upsert flow verified (dataset ${JSON.stringify(name)})`);
}

const client = createClient();

void main().catch((error: unknown) => {
  console.error(error);
  process.exit(1);
});
